Add support for end-to-end functional/ui testing

webteckie opened the pull request:
Basic support for nightwatchjs testing to guard against functional/ui regressions.

ericelliott commented on test/e2e/README.md:
```diff
@@ -0,0 +1,22 @@
+        
+## Running
+        
+    npm run-script test-e2e
```
This can be shortened to `npm run test-e2e`.

ericelliott commented:
Looks like a good start. We also need to add it to CI. Do we have a Sauce Labs account for KeystoneJS? If not, we should set one up. =)

What's going on with the broken build?

webteckie commented:
Hmm...looks like mocha is picking /home/travis/build/keystonejs/keystone/test/e2e/server.js as one of its tests.  I need to exclude the e2e directory from mocha or move it to outside the test directory onto its own top level directory.

webteckie commented:
I have separated the tests into their own subdirectories:  `test\unit_tests` and `test\e2e`

Not sure why Travis is failing the following...it works fine on my system:
```
1) List "track" option "before all" hook:
 Uncaught TypeError: err is not a function
```

And have no clue on the final errors:
```
npm ERR! Linux 3.13.0-40-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v4.0.0/bin/node" "/home/travis/.nvm/versions/node/v4.0.0/bin/npm" "run" "test-cov"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha test-cov: `istanbul cover -x 'test/e2e/**' ./node_modules/mocha/bin/_mocha`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha test-cov script 'istanbul cover -x 'test/e2e/**' ./node_modules/mocha/bin/_mocha'.
npm ERR! This is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     istanbul cover -x 'test/e2e/**' ./node_modules/mocha/bin/_mocha
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/keystonejs/keystone/npm-debug.log
```

If anyone has any ideas please let me know!

webteckie commented:
I'm stumped...if anyone has a clue as to why the track test is failing let me know...otherwise, I'll continue with this tomorrow evening.

ericelliott commented:
Sure. Not sure all those "unit tests" are actually unit tests, but `¯\_(ツ)_/¯`

Maybe it should be `unitTests` instead of `unit_tests` to follow the convention of the rest of the repo?

Personally I prefer dashes. Some unix-based filesystems are case sensitive. Others are case-insensitive, so I don't think camelCase is the best idea for cross-OS compatibility.

My $.02. Feel free to ignore. =)

morenoh149 commented:
I took a stab at running your PR @webteckie but I get an unrelated error
```
npm run test-cov

> keystone@0.4.0-alpha pretest-cov /Users/harrymoreno/programming/keystone/keystone
> npm run clean && npm run lint


> keystone@0.4.0-alpha clean /Users/harrymoreno/programming/keystone/keystone
> rimraf ./coverage

module.js:328
    throw err;
    ^

Error: Cannot find module 'minimatch'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/node_modules/rimraf/node_modules/glob/glob.js:44:17)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)

npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/node" "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/npm" "run" "clean"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha clean: `rimraf ./coverage`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha clean script 'rimraf ./coverage'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     rimraf ./coverage
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/harrymoreno/programming/keystone/keystone/npm-debug.log

npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/node" "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/npm" "run" "test-cov"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha pretest-cov: `npm run clean && npm run lint`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha pretest-cov script 'npm run clean && npm run lint'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run clean && npm run lint
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/harrymoreno/programming/keystone/keystone/npm-debug.log
```
using node v5.4.0 and npm 3.3.12

morenoh149 commented:
npm lint
```
 npm run clean && npm run lint

> keystone@0.4.0-alpha clean /Users/harrymoreno/programming/keystone/keystone
> rimraf ./coverage

module.js:328
    throw err;
    ^

Error: Cannot find module 'minimatch'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/node_modules/rimraf/node_modules/glob/glob.js:44:17)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)

npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/node" "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/npm" "run" "clean"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha clean: `rimraf ./coverage`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha clean script 'rimraf ./coverage'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     rimraf ./coverage
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/harrymoreno/programming/keystone/keystone/npm-debug.log
```

morenoh149 commented:
minimatch should be added to the devdependencies? I get some eslint errors after `npm install minimatch` in the directory

webteckie commented:
@morenoh149 did you `npm install` under keystone before trying to run the tests?  I didn't get those errors.

morenoh149 commented:
I think I reproduced the travis error after rm -rfing node_modules and npm install
```
 npm run test-cov

> keystone@0.4.0-alpha pretest-cov /Users/harrymoreno/programming/keystone/keystone
> npm run clean && npm run lint


> keystone@0.4.0-alpha clean /Users/harrymoreno/programming/keystone/keystone
> rimraf ./coverage


> keystone@0.4.0-alpha lint /Users/harrymoreno/programming/keystone/keystone
> eslint .


> keystone@0.4.0-alpha test-cov /Users/harrymoreno/programming/keystone/keystone
> istanbul cover ./node_modules/mocha/bin/_mocha

Error: Trying to open unclosed connection.
    at NativeConnection.Connection.open (/Users/harrymoreno/programming/keystone/keystone/node_modules/mongoose/lib/connection.js:235:15)
    at Mongoose.connect (/Users/harrymoreno/programming/keystone/keystone/node_modules/mongoose/lib/index.js:240:31)
    at openDatabaseConnection (/Users/harrymoreno/programming/keystone/keystone/lib/core/openDatabaseConnection.js:9:1822)
    at start (/Users/harrymoreno/programming/keystone/keystone/lib/core/start.js:9:1878)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/test/e2e/server.js:28:10)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:219:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:216:14)
    at Mocha.run (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:468:10)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:430:10)
    at runFn (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/command/common/run-with-cover.js:122:16)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/command/common/run-with-cover.js:251:17
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/util/file-matcher.js:68:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:52:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:361:13
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:52:16
    at done (/Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:246:17)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:44:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:358:17
    at LOOP (fs.js:1613:14)
    at nextTickCallbackWith0Args (node.js:456:9)
    at process._tickCallback (node.js:385:13)
=============================================================================
Writing coverage object [/Users/harrymoreno/programming/keystone/keystone/coverage/coverage.json]
Writing coverage reports at [/Users/harrymoreno/programming/keystone/keystone/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 32.08% ( 1293/4031 )
Branches     : 10.93% ( 288/2636 )
Functions    : 15.59% ( 97/622 )
Lines        : 32.95% ( 1291/3918 )
================================================================================

npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/node" "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/npm" "run" "test-cov"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha test-cov: `istanbul cover ./node_modules/mocha/bin/_mocha`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha test-cov script 'istanbul cover ./node_modules/mocha/bin/_mocha'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     istanbul cover ./node_modules/mocha/bin/_mocha
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/harrymoreno/programming/keystone/keystone/npm-debug.log
```

webteckie commented:
@ericelliott i'm not a fan of mix casing file/directory names for that reason that some unix OSs are case sensitive and thus it makes it easier to normalize to lowercase.   But if there's already a guideline that keystone uses camelCase for everything then I should comply.  Will address that tomorrow =)

morenoh149 commented:
Got a very similar error to the one travis reports
```
➜   npm run test-cov

> keystone@0.4.0-alpha pretest-cov /Users/harrymoreno/programming/keystone/keystone
> npm run clean && npm run lint


> keystone@0.4.0-alpha clean /Users/harrymoreno/programming/keystone/keystone
> rimraf ./coverage


> keystone@0.4.0-alpha lint /Users/harrymoreno/programming/keystone/keystone
> eslint .


> keystone@0.4.0-alpha test-cov /Users/harrymoreno/programming/keystone/keystone
> istanbul cover ./node_modules/mocha/bin/_mocha

Error: Trying to open unclosed connection.
    at NativeConnection.Connection.open (/Users/harrymoreno/programming/keystone/keystone/node_modules/mongoose/lib/connection.js:235:15)
    at Mongoose.connect (/Users/harrymoreno/programming/keystone/keystone/node_modules/mongoose/lib/index.js:240:31)
    at openDatabaseConnection (/Users/harrymoreno/programming/keystone/keystone/lib/core/openDatabaseConnection.js:9:1822)
    at start (/Users/harrymoreno/programming/keystone/keystone/lib/core/start.js:9:1878)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/test/e2e/server.js:28:10)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:219:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:216:14)
    at Mocha.run (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/lib/mocha.js:468:10)
    at Object.<anonymous> (/Users/harrymoreno/programming/keystone/keystone/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:430:10)
    at runFn (/Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/command/common/run-with-cover.js:122:16)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/command/common/run-with-cover.js:251:17
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/istanbul/lib/util/file-matcher.js:68:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:52:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:361:13
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:52:16
    at done (/Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:246:17)
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:44:16
    at /Users/harrymoreno/programming/keystone/keystone/node_modules/async/lib/async.js:358:17
    at LOOP (fs.js:1613:14)
    at nextTickCallbackWith0Args (node.js:456:9)
    at process._tickCallback (node.js:385:13)
=============================================================================
Writing coverage object [/Users/harrymoreno/programming/keystone/keystone/coverage/coverage.json]
Writing coverage reports at [/Users/harrymoreno/programming/keystone/keystone/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 32.08% ( 1293/4031 )
Branches     : 10.93% ( 288/2636 )
Functions    : 15.59% ( 97/622 )
Lines        : 32.95% ( 1291/3918 )
================================================================================

npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/node" "/Users/harrymoreno/.nvm/versions/node/v5.4.0/bin/npm" "run" "test-cov"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! keystone@0.4.0-alpha test-cov: `istanbul cover ./node_modules/mocha/bin/_mocha`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the keystone@0.4.0-alpha test-cov script 'istanbul cover ./node_modules/mocha/bin/_mocha'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the keystone package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     istanbul cover ./node_modules/mocha/bin/_mocha
npm ERR! You can get their info via:
npm ERR!     npm owner ls keystone
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/harrymoreno/programming/keystone/keystone/npm-debug.log
```
sometime after the e2e server starts https://github.com/keystonejs/keystone/pull/2254/files#diff-1a4e05a3b3fe0ce6e9820094a87abf88R28

morenoh149 commented on package.json:
```diff
@@ -145,6 +146,7 @@
     "release": "gulp release",
     "pretest": "npm run lint",
     "test": "mocha",
+    "test-e2e": "node test/e2e/server.js --config ./test/e2e/nightwatch.json",
```
this could be prepended with `java -jar bin/selenium-server-standalone-2.51.0.jar`

morenoh149 commented on test/e2e/test_groups/login_group/login_tests.js:
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  'Keystone should allow admins to login to the Admin UI' : function (client) {
+    client
+      .url('http://XPS8500:3000/keystone')
```
shouldn't this be localhost:3000 ?

ericelliott replied:
:+1:

webteckie replied:
I only wanted it to work in my machine...crap=)  Hopefully will have time to address all the comments this evening (EST).

morenoh149 commented:
mocha probably shouldn't be testing the nightwatch script. I think that's the error above.

morenoh149 commented:
All occurrences of `throw new err` should probably be changed to `throw new Error(err)` or something syntactically correct as `err` is not necessarily a function.

https://github.com/keystonejs/keystone/blob/cc4385037151f6c1e3af5dee342672fb6b8e6954/test/track.test.js#L116

webteckie commented:
Alright...took care of the particular track test failures and addressed all code review comments.

@JedWatson if this all looks good to you I think we should merge now before any other work continues given that the test directories were rearranged to fit in the e2e testing.

ericelliott commented:
:+1:

JedWatson commented:
@webteckie looks great - thanks, and cheers everyone for the code reviews along the way!

@morenoh149 `lint-fix` is a little helper task I added while working through the linter errors. It just runs eslint with the `--fix` flag, which tells eslint to automatically fix errors it can (e.g. trailing whitespace)

Happy to remove it again but it's much nicer than typing `node_modules/bin/eslint . --fix` to achieve this so I added it for now. Ambivalent :)

@ericelliott @webteckie re: directory casing, `camelCase` seems to be a pretty strong standard among node/react repos, most files that export a single method or "Class" (`React.createClass`) have a filename with matching case. I personally think using `camelCase` or `ClassCase` (or whatever your preferred names for those cases are) for methods and `kebab-case` for file & directory names would be a bit weird but if it causes a problem down the track I'm happy to review it.

There's an exception in the public `js` folder where directories are named after their packages (e.g. `query-placeholder-shim`) but again that's for consistency between the name of the file / folder and what it contains.
